File: server/storage.ts
import { 
  users, 
  tasks, 
  studySessions, 
  notes, 
  flashcardSets, 
  flashcards, 
  studyProgress,
  type User, 
  type InsertUser,
  type Task,
  type InsertTask,
  type StudySession,
  type InsertStudySession,
  type Note,
  type InsertNote,
  type FlashcardSet,
  type InsertFlashcardSet,
  type Flashcard,
  type InsertFlashcard,
  type StudyProgress,
  type InsertStudyProgress
} from "@shared/schema";

// Interface for storage operations
export interface IStorage {
  // User operations
  getUser(id: number): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Task operations
  getTasks(userId: number): Promise<Task[]>;
  getTaskById(id: number): Promise<Task | undefined>;
  createTask(task: InsertTask): Promise<Task>;
  updateTask(id: number, task: Partial<Task>): Promise<Task | undefined>;
  deleteTask(id: number): Promise<boolean>;
  
  // Study session operations
  getStudySessions(userId: number): Promise<StudySession[]>;
  getStudySessionById(id: number): Promise<StudySession | undefined>;
  createStudySession(session: InsertStudySession): Promise<StudySession>;
  updateStudySession(id: number, session: Partial<StudySession>): Promise<StudySession | undefined>;
  deleteStudySession(id: number): Promise<boolean>;
  
  // Note operations
  getNotes(userId: number): Promise<Note[]>;
  getNoteById(id: number): Promise<Note | undefined>;
  createNote(note: InsertNote): Promise<Note>;
  updateNote(id: number, note: Partial<Note>): Promise<Note | undefined>;
  deleteNote(id: number): Promise<boolean>;
  
  // Flashcard set operations
  getFlashcardSets(userId: number): Promise<FlashcardSet[]>;
  getFlashcardSetById(id: number): Promise<FlashcardSet | undefined>;
  createFlashcardSet(set: InsertFlashcardSet): Promise<FlashcardSet>;
  updateFlashcardSet(id: number, set: Partial<FlashcardSet>): Promise<FlashcardSet | undefined>;
  deleteFlashcardSet(id: number): Promise<boolean>;
  
  // Flashcard operations
  getFlashcards(setId: number): Promise<Flashcard[]>;
  getFlashcardById(id: number): Promise<Flashcard | undefined>;
  createFlashcard(flashcard: InsertFlashcard): Promise<Flashcard>;
  updateFlashcard(id: number, flashcard: Partial<Flashcard>): Promise<Flashcard | undefined>;
  deleteFlashcard(id: number): Promise<boolean>;
  
  // Study progress operations
  getStudyProgress(userId: number): Promise<StudyProgress[]>;
  createStudyProgress(progress: InsertStudyProgress): Promise<StudyProgress>;
}

export class MemStorage implements IStorage {
  private users: Map<number, User>;
  private tasks: Map<number, Task>;
  private studySessions: Map<number, StudySession>;
  private notes: Map<number, Note>;
  private flashcardSets: Map<number, FlashcardSet>;
  private flashcards: Map<number, Flashcard>;
  private studyProgress: Map<number, StudyProgress>;
  
  // ID counters for each entity
  private userIdCounter: number;
  private taskIdCounter: number;
  private sessionIdCounter: number;
  private noteIdCounter: number;
  private setIdCounter: number;
  private flashcardIdCounter: number;
  private progressIdCounter: number;

  constructor() {
    this.users = new Map();
    this.tasks = new Map();
    this.studySessions = new Map();
    this.notes = new Map();
    this.flashcardSets = new Map();
    this.flashcards = new Map();
    this.studyProgress = new Map();
    
    this.userIdCounter = 1;
    this.taskIdCounter = 1;
    this.sessionIdCounter = 1;
    this.noteIdCounter = 1;
    this.setIdCounter = 1;
    this.flashcardIdCounter = 1;
    this.progressIdCounter = 1;
    
    // Add a default user
    this.createUser({
      username: "alexjohnson",
      password: "password123",
      name: "Alex Johnson",
      email: "alex@example.com",
      major: "Computer Science",
      avatarUrl: ""
    });
  }

  // User operations
  async getUser(id: number): Promise<User | undefined> {
    return this.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (user) => user.username === username,
    );
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = this.userIdCounter++;
    const user: User = { ...insertUser, id };
    this.users.set(id, user);
    return user;
  }

  // Task operations
  async getTasks(userId: number): Promise<Task[]> {
    return Array.from(this.tasks.values()).filter(
      (task) => task.userId === userId,
    );
  }

  async getTaskById(id: number): Promise<Task | undefined> {
    return this.tasks.get(id);
  }

  async createTask(task: InsertTask): Promise<Task> {
    const id = this.taskIdCounter++;
    const newTask: Task = { ...task, id };
    this.tasks.set(id, newTask);
    return newTask;
  }

  async updateTask(id: number, taskUpdate: Partial<Task>): Promise<Task | undefined> {
    const task = this.tasks.get(id);
    if (!task) return undefined;
    
    const updatedTask = { ...task, ...taskUpdate };
    this.tasks.set(id, updatedTask);
    return updatedTask;
  }

  async deleteTask(id: number): Promise<boolean> {
    return this.tasks.delete(id);
  }

  // Study session operations
  async getStudySessions(userId: number): Promise<StudySession[]> {
    return Array.from(this.studySessions.values()).filter(
      (session) => session.userId === userId,
    );
  }

  async getStudySessionById(id: number): Promise<StudySession | undefined> {
    return this.studySessions.get(id);
  }

  async createStudySession(session: InsertStudySession): Promise<StudySession> {
    const id = this.sessionIdCounter++;
    const newSession: StudySession = { ...session, id };
    this.studySessions.set(id, newSession);
    return newSession;
  }

  async updateStudySession(id: number, sessionUpdate: Partial<StudySession>): Promise<StudySession | undefined> {
    const session = this.studySessions.get(id);
    if (!session) return undefined;
    
    const updatedSession = { ...session, ...sessionUpdate };
    this.studySessions.set(id, updatedSession);
    return updatedSession;
  }

  async deleteStudySession(id: number): Promise<boolean> {
    return this.studySessions.delete(id);
  }

  // Note operations
  async getNotes(userId: number): Promise<Note[]> {
    return Array.from(this.notes.values()).filter(
      (note) => note.userId === userId,
    );
  }

  async getNoteById(id: number): Promise<Note | undefined> {
    return this.notes.get(id);
  }

  async createNote(note: InsertNote): Promise<Note> {
    const id = this.noteIdCounter++;
    const now = new Date();
    const newNote: Note = { 
      ...note, 
      id, 
      createdAt: now, 
      updatedAt: now 
    };
    this.notes.set(id, newNote);
    return newNote;
  }

  async updateNote(id: number, noteUpdate: Partial<Note>): Promise<Note | undefined> {
    const note = this.notes.get(id);
    if (!note) return undefined;
    
    const updatedNote = { 
      ...note, 
      ...noteUpdate,
      updatedAt: new Date()
    };
    this.notes.set(id, updatedNote);
    return updatedNote;
  }

  async deleteNote(id: number): Promise<boolean> {
    return this.notes.delete(id);
  }

  // Flashcard set operations
  async getFlashcardSets(userId: number): Promise<FlashcardSet[]> {
    return Array.from(this.flashcardSets.values()).filter(
      (set) => set.userId === userId,
    );
  }

  async getFlashcardSetById(id: number): Promise<FlashcardSet | undefined> {
    return this.flashcardSets.get(id);
  }

  async createFlashcardSet(set: InsertFlashcardSet): Promise<FlashcardSet> {
    const id = this.setIdCounter++;
    const now = new Date();
    const newSet: FlashcardSet = { 
      ...set, 
      id, 
      createdAt: now
    };
    this.flashcardSets.set(id, newSet);
    return newSet;
  }

  async updateFlashcardSet(id: number, setUpdate: Partial<FlashcardSet>): Promise<FlashcardSet | undefined> {
    const set = this.flashcardSets.get(id);
    if (!set) return undefined;
    
    const updatedSet = { 
      ...set, 
      ...setUpdate
    };
    this.flashcardSets.set(id, updatedSet);
    return updatedSet;
  }

  async deleteFlashcardSet(id: number): Promise<boolean> {
    // Also delete all flashcards belonging to this set
    Array.from(this.flashcards.values())
      .filter(card => card.setId === id)
      .forEach(card => this.flashcards.delete(card.id));
    
    return this.flashcardSets.delete(id);
  }

  // Flashcard operations
  async getFlashcards(setId: number): Promise<Flashcard[]> {
    return Array.from(this.flashcards.values()).filter(
      (card) => card.setId === setId,
    );
  }

  async getFlashcardById(id: number): Promise<Flashcard | undefined> {
    return this.flashcards.get(id);
  }

  async createFlashcard(flashcard: InsertFlashcard): Promise<Flashcard> {
    const id = this.flashcardIdCounter++;
    const newCard: Flashcard = { ...flashcard, id };
    this.flashcards.set(id, newCard);
    return newCard;
  }

  async updateFlashcard(id: number, cardUpdate: Partial<Flashcard>): Promise<Flashcard | undefined> {
    const card = this.flashcards.get(id);
    if (!card) return undefined;
    
    const updatedCard = { ...card, ...cardUpdate };
    this.flashcards.set(id, updatedCard);
    return updatedCard;
  }

  async deleteFlashcard(id: number): Promise<boolean> {
    return this.flashcards.delete(id);
  }

  // Study progress operations
  async getStudyProgress(userId: number): Promise<StudyProgress[]> {
    return Array.from(this.studyProgress.values()).filter(
      (progress) => progress.userId === userId,
    );
  }

  async createStudyProgress(progress: InsertStudyProgress): Promise<StudyProgress> {
    const id = this.progressIdCounter++;
    const newProgress: StudyProgress = { ...progress, id };
    this.studyProgress.set(id, newProgress);
    return newProgress;
  }
}

// Create and export a shared instance of the storage
export const storage = new MemStorage();

// Initialize with sample data
const initializeStorage = async () => {
  const user = await storage.getUserByUsername("alexjohnson");
  if (!user) return;
  
  // Create demo tasks
  await storage.createTask({
    userId: user.id,
    title: "Complete ML Project",
    description: "Finalize the machine learning project with classification model",
    dueDate: new Date(Date.now() + 24 * 60 * 60 * 1000), // tomorrow
    priority: 1, // high
    completed: false,
    category: "Projects"
  });
  
  await storage.createTask({
    userId: user.id,
    title: "Study for Database Exam",
    description: "Review SQL queries, normalization, and indexing",
    dueDate: new Date(Date.now() + 3 * 24 * 60 * 60 * 1000), // 3 days
    priority: 2, // medium
    completed: false,
    category: "Exams"
  });
  
  await storage.createTask({
    userId: user.id,
    title: "Review Lecture Notes",
    description: "Go through the notes from last week's lectures",
    dueDate: new Date(Date.now() + 5 * 24 * 60 * 60 * 1000), // 5 days
    priority: 3, // low
    completed: false,
    category: "Review"
  });
  
  await storage.createTask({
    userId: user.id,
    title: "Research Paper Outline",
    description: "Create an outline for the upcoming research paper",
    dueDate: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000), // 7 days
    priority: 3, // low
    completed: false,
    category: "Papers"
  });
  
  // Create study sessions
  const today = new Date();
  
  await storage.createStudySession({
    userId: user.id,
    title: "Data Structures Lecture",
    startTime: new Date(today.setHours(9, 0, 0, 0)),
    endTime: new Date(today.setHours(10, 30, 0, 0)),
    subject: "Computer Science",
    description: "Lecture on advanced data structures",
    location: "Room 302, Engineering Building"
  });
  
  await storage.createStudySession({
    userId: user.id,
    title: "Study Group - Algorithms",
    startTime: new Date(today.setHours(11, 30, 0, 0)),
    endTime: new Date(today.setHours(13, 0, 0, 0)),
    subject: "Computer Science",
    description: "Group study for algorithm optimization",
    location: "Library Study Room 4"
  });
  
  await storage.createStudySession({
    userId: user.id,
    title: "Assignment Review - AI Ethics",
    startTime: new Date(today.setHours(14, 0, 0, 0)),
    endTime: new Date(today.setHours(15, 0, 0, 0)),
    subject: "AI Ethics",
    description: "Review session for the AI ethics assignment",
    location: "Zoom Meeting"
  });
  
  // Create flashcard sets
  const dsaSet = await storage.createFlashcardSet({
    userId: user.id,
    title: "Data Structures & Algorithms",
    description: "Flashcards for DSA course",
    subject: "Computer Science",
    tags: ["algorithms", "data structures", "complexity"]
  });
  
  const archSet = await storage.createFlashcardSet({
    userId: user.id,
    title: "Computer Architecture",
    description: "Flashcards for Computer Architecture class",
    subject: "Computer Science",
    tags: ["cpu", "memory", "architecture"]
  });
  
  // Create flashcards
  await storage.createFlashcard({
    setId: dsaSet.id,
    question: "What is the time complexity of QuickSort in the worst case?",
    answer: "O(n²) - This occurs when the pivot selection is poor, such as when the array is already sorted.",
    proficiency: 2
  });
  
  await storage.createFlashcard({
    setId: archSet.id,
    question: "What are the four main components of a CPU?",
    answer: "1. Arithmetic Logic Unit (ALU)\n2. Control Unit\n3. Registers\n4. Cache",
    proficiency: 3
  });
  
  // Create study progress entries
  const sevenDaysAgo = new Date();
  sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 6);
  
  for (let i = 0; i < 7; i++) {
    const date = new Date(sevenDaysAgo);
    date.setDate(date.getDate() + i);
    
    // Randomize study duration between 30 and 240 minutes
    const minutes = Math.floor(Math.random() * 211) + 30;
    
    await storage.createStudyProgress({
      userId: user.id,
      date,
      studyDuration: minutes,
      subject: "Various",
      notes: `Study session on day ${i+1}`
    });
  }
};

// Initialize the storage with sample data
initializeStorage();

File: python_server/storage.py
from datetime import datetime
from copy import deepcopy
from typing import Dict, List, Optional, Any, TypedDict, Union

# Type definitions
class User(TypedDict):
    id: int
    username: str
    password: str
    name: str
    email: str
    major: Optional[str]
    avatarUrl: Optional[str]

class InsertUser(TypedDict):
    username: str
    password: str
    name: str
    email: str
    major: Optional[str]
    avatarUrl: Optional[str]

class Task(TypedDict):
    id: int
    userId: int
    title: str
    description: Optional[str]
    dueDate: Optional[str]
    priority: int  # 1 = high, 2 = medium, 3 = low
    completed: bool
    category: Optional[str]

class InsertTask(TypedDict):
    userId: int
    title: str
    description: Optional[str]
    dueDate: Optional[str]
    priority: int
    completed: bool
    category: Optional[str]

class StudySession(TypedDict):
    id: int
    userId: int
    title: str
    startTime: str
    endTime: Optional[str]
    subject: Optional[str]
    description: Optional[str]
    location: Optional[str]

class InsertStudySession(TypedDict):
    userId: int
    title: str
    startTime: str
    endTime: Optional[str]
    subject: Optional[str]
    description: Optional[str]
    location: Optional[str]

class Note(TypedDict):
    id: int
    userId: int
    title: str
    content: str
    subject: Optional[str]
    tags: Optional[List[str]]
    createdAt: Union[str, datetime]
    updatedAt: Union[str, datetime]

class InsertNote(TypedDict):
    userId: int
    title: str
    content: str
    subject: Optional[str]
    tags: Optional[List[str]]

class FlashcardSet(TypedDict):
    id: int
    userId: int
    title: str
    description: Optional[str]
    subject: Optional[str]
    tags: Optional[List[str]]
    createdAt: Union[str, datetime]

class InsertFlashcardSet(TypedDict):
    userId: int
    title: str
    description: Optional[str]
    subject: Optional[str]
    tags: Optional[List[str]]

class Flashcard(TypedDict):
    id: int
    setId: int
    question: str
    answer: str
    lastReviewed: Optional[str]
    proficiency: Optional[int]  # 0-4 scale for spaced repetition

class InsertFlashcard(TypedDict):
    setId: int
    question: str
    answer: str
    lastReviewed: Optional[str]
    proficiency: Optional[int]

class StudyProgress(TypedDict):
    id: int
    userId: int
    date: Union[str, datetime]
    studyDuration: int  # in minutes
    subject: Optional[str]
    notes: Optional[str]

class InsertStudyProgress(TypedDict):
    userId: int
    date: Union[str, datetime]
    studyDuration: int
    subject: Optional[str]
    notes: Optional[str]

# Interface for storage operations
class IStorage:
    # User operations
    def get_user(self, id: int) -> Optional[User]: pass
    def get_user_by_username(self, username: str) -> Optional[User]: pass
    def create_user(self, user: InsertUser) -> User: pass
    
    # Task operations
    def get_tasks(self, user_id: int) -> List[Task]: pass
    def get_task_by_id(self, id: int) -> Optional[Task]: pass
    def create_task(self, task: InsertTask) -> Task: pass
    def update_task(self, id: int, task: Dict[str, Any]) -> Optional[Task]: pass
    def delete_task(self, id: int) -> bool: pass
    
    # Study session operations
    def get_study_sessions(self, user_id: int) -> List[StudySession]: pass
    def get_study_session_by_id(self, id: int) -> Optional[StudySession]: pass
    def create_study_session(self, session: InsertStudySession) -> StudySession: pass
    def update_study_session(self, id: int, session: Dict[str, Any]) -> Optional[StudySession]: pass
    def delete_study_session(self, id: int) -> bool: pass
    
    # Note operations
    def get_notes(self, user_id: int) -> List[Note]: pass
    def get_note_by_id(self, id: int) -> Optional[Note]: pass
    def create_note(self, note: InsertNote) -> Note: pass
    def update_note(self, id: int, note: Dict[str, Any]) -> Optional[Note]: pass
    def delete_note(self, id: int) -> bool: pass
    
    # Flashcard set operations
    def get_flashcard_sets(self, user_id: int) -> List[FlashcardSet]: pass
    def get_flashcard_set_by_id(self, id: int) -> Optional[FlashcardSet]: pass
    def create_flashcard_set(self, set: InsertFlashcardSet) -> FlashcardSet: pass
    def update_flashcard_set(self, id: int, set: Dict[str, Any]) -> Optional[FlashcardSet]: pass
    def delete_flashcard_set(self, id: int) -> bool: pass
    
    # Flashcard operations
    def get_flashcards(self, set_id: int) -> List[Flashcard]: pass
    def get_flashcard_by_id(self, id: int) -> Optional[Flashcard]: pass
    def create_flashcard(self, flashcard: InsertFlashcard) -> Flashcard: pass
    def update_flashcard(self, id: int, flashcard: Dict[str, Any]) -> Optional[Flashcard]: pass
    def delete_flashcard(self, id: int) -> bool: pass
    
    # Study progress operations
    def get_study_progress(self, user_id: int) -> List[StudyProgress]: pass
    def create_study_progress(self, progress: InsertStudyProgress) -> StudyProgress: pass

class MemStorage(IStorage):
    def __init__(self):
        self.users: Dict[int, User] = {}
        self.tasks: Dict[int, Task] = {}
        self.study_sessions: Dict[int, StudySession] = {}
        self.notes: Dict[int, Note] = {}
        self.flashcard_sets: Dict[int, FlashcardSet] = {}
        self.flashcards: Dict[int, Flashcard] = {}
        self.study_progress: Dict[int, StudyProgress] = {}
        
        # ID counters for each entity
        self.user_id_counter = 1
        self.task_id_counter = 1
        self.session_id_counter = 1
        self.note_id_counter = 1
        self.set_id_counter = 1
        self.flashcard_id_counter = 1
        self.progress_id_counter = 1
        
        # Add a default user
        self.create_user({
            "username": "alexjohnson",
            "password": "password123",
            "name": "Alex Johnson",
            "email": "alex@example.com",
            "major": "Computer Science",
            "avatarUrl": ""
        })

    # User operations
    def get_user(self, id: int) -> Optional[User]:
        return self.users.get(id)

    def get_user_by_username(self, username: str) -> Optional[User]:
        for user in self.users.values():
            if user["username"] == username:
                return user
        return None

    def create_user(self, user: InsertUser) -> User:
        id = self.user_id_counter
        self.user_id_counter += 1
        new_user: User = {**user, "id": id}
        self.users[id] = new_user
        return new_user

    # Task operations
    def get_tasks(self, user_id: int) -> List[Task]:
        return [task for task in self.tasks.values() if task["userId"] == user_id]

    def get_task_by_id(self, id: int) -> Optional[Task]:
        return self.tasks.get(id)

    def create_task(self, task: InsertTask) -> Task:
        id = self.task_id_counter
        self.task_id_counter += 1
        new_task: Task = {**task, "id": id}
        self.tasks[id] = new_task
        return new_task

    def update_task(self, id: int, task_update: Dict[str, Any]) -> Optional[Task]:
        task = self.tasks.get(id)
        if not task:
            return None
        
        updated_task = {**task, **task_update}
        self.tasks[id] = updated_task
        return updated_task

    def delete_task(self, id: int) -> bool:
        if id in self.tasks:
            del self.tasks[id]
            return True
        return False

    # Study session operations
    def get_study_sessions(self, user_id: int) -> List[StudySession]:
        return [session for session in self.study_sessions.values() if session["userId"] == user_id]

    def get_study_session_by_id(self, id: int) -> Optional[StudySession]:
        return self.study_sessions.get(id)

    def create_study_session(self, session: InsertStudySession) -> StudySession:
        id = self.session_id_counter
        self.session_id_counter += 1
        new_session: StudySession = {**session, "id": id}
        self.study_sessions[id] = new_session
        return new_session

    def update_study_session(self, id: int, session_update: Dict[str, Any]) -> Optional[StudySession]:
        session = self.study_sessions.get(id)
        if not session:
            return None
        
        updated_session = {**session, **session_update}
        self.study_sessions[id] = updated_session
        return updated_session

    def delete_study_session(self, id: int) -> bool:
        if id in self.study_sessions:
            del self.study_sessions[id]
            return True
        return False

    # Note operations
    def get_notes(self, user_id: int) -> List[Note]:
        return [note for note in self.notes.values() if note["userId"] == user_id]

    def get_note_by_id(self, id: int) -> Optional[Note]:
        return self.notes.get(id)

    def create_note(self, note: InsertNote) -> Note:
        id = self.note_id_counter
        self.note_id_counter += 1
        now = datetime.now().isoformat()
        new_note: Note = {
            **note,
            "id": id,
            "createdAt": now,
            "updatedAt": now
        }
        self.notes[id] = new_note
        return new_note

    def update_note(self, id: int, note_update: Dict[str, Any]) -> Optional[Note]:
        note = self.notes.get(id)
        if not note:
            return None
        
        updated_note = {
            **note,
            **note_update,
            "updatedAt": datetime.now().isoformat()
        }
        self.notes[id] = updated_note
        return updated_note

    def delete_note(self, id: int) -> bool:
        if id in self.notes:
            del self.notes[id]
            return True
        return False

    # Flashcard set operations
    def get_flashcard_sets(self, user_id: int) -> List[FlashcardSet]:
        return [set for set in self.flashcard_sets.values() if set["userId"] == user_id]

    def get_flashcard_set_by_id(self, id: int) -> Optional[FlashcardSet]:
        return self.flashcard_sets.get(id)

    def create_flashcard_set(self, set: InsertFlashcardSet) -> FlashcardSet:
        id = self.set_id_counter
        self.set_id_counter += 1
        now = datetime.now().isoformat()
        new_set: FlashcardSet = {
            **set,
            "id": id,
            "createdAt": now
        }
        self.flashcard_sets[id] = new_set
        return new_set

    def update_flashcard_set(self, id: int, set_update: Dict[str, Any]) -> Optional[FlashcardSet]:
        set_data = self.flashcard_sets.get(id)
        if not set_data:
            return None
        
        updated_set = {**set_data, **set_update}
        self.flashcard_sets[id] = updated_set
        return updated_set

    def delete_flashcard_set(self, id: int) -> bool:
        # Also delete all flashcards belonging to this set
        flashcards_to_delete = [card_id for card_id, card in self.flashcards.items() if card["setId"] == id]
        for card_id in flashcards_to_delete:
            del self.flashcards[card_id]
        
        if id in self.flashcard_sets:
            del self.flashcard_sets[id]
            return True
        return False

    # Flashcard operations
    def get_flashcards(self, set_id: int) -> List[Flashcard]:
        return [card for card in self.flashcards.values() if card["setId"] == set_id]

    def get_flashcard_by_id(self, id: int) -> Optional[Flashcard]:
        return self.flashcards.get(id)

    def create_flashcard(self, flashcard: InsertFlashcard) -> Flashcard:
        id = self.flashcard_id_counter
        self.flashcard_id_counter += 1
        new_card: Flashcard = {**flashcard, "id": id}
        self.flashcards[id] = new_card
        return new_card

    def update_flashcard(self, id: int, card_update: Dict[str, Any]) -> Optional[Flashcard]:
        card = self.flashcards.get(id)
        if not card:
            return None
        
        updated_card = {**card, **card_update}
        self.flashcards[id] = updated_card
        return updated_card

    def delete_flashcard(self, id: int) -> bool:
        if id in self.flashcards:
            del self.flashcards[id]
            return True
        return False

    # Study progress operations
    def get_study_progress(self, user_id: int) -> List[StudyProgress]:
        return [progress for progress in self.study_progress.values() if progress["userId"] == user_id]

    def create_study_progress(self, progress: InsertStudyProgress) -> StudyProgress:
        id = self.progress_id_counter
        self.progress_id_counter += 1
        new_progress: StudyProgress = {**progress, "id": id}
        self.study_progress[id] = new_progress
        return new_progress

# Create and export a shared instance of the storage
storage = MemStorage()

# Initialize with sample data
def initialize_storage():
    user = storage.get_user_by_username("alexjohnson")
    if not user:
        return
    
    # Create demo tasks
    storage.create_task({
        "userId": user["id"],
        "title": "Complete ML Project",
        "description": "Finalize the machine learning project with classification model",
        "dueDate": (datetime.now().timestamp() + 24 * 60 * 60) * 1000,  # tomorrow
        "priority": 1,  # high
        "completed": False,
        "category": "Projects"
    })
    
    storage.create_task({
        "userId": user["id"],
        "title": "Study for Database Exam",
        "description": "Review SQL queries, normalization, and indexing",
        "dueDate": (datetime.now().timestamp() + 3 * 24 * 60 * 60) * 1000,  # 3 days
        "priority": 2,  # medium
        "completed": False,
        "category": "Exams"
    })
    
    storage.create_task({
        "userId": user["id"],
        "title": "Review Lecture Notes",
        "description": "Go through the notes from last week's lectures",
        "dueDate": (datetime.now().timestamp() + 5 * 24 * 60 * 60) * 1000,  # 5 days
        "priority": 3,  # low
        "completed": False,
        "category": "Review"
    })
    
    storage.create_task({
        "userId": user["id"],
        "title": "Research Paper Outline",
        "description": "Create an outline for the upcoming research paper",
        "dueDate": (datetime.now().timestamp() + 7 * 24 * 60 * 60) * 1000,  # 7 days
        "priority": 3,  # low
        "completed": False,
        "category": "Papers"
    })
    
    # Create study sessions
    today = datetime.now().replace(hour=9, minute=0, second=0, microsecond=0)
    
    storage.create_study_session({
        "userId": user["id"],
        "title": "Data Structures Lecture",
        "startTime": today.isoformat(),
        "endTime": today.replace(hour=10, minute=30).isoformat(),
        "subject": "Computer Science",
        "description": "Lecture on advanced data structures",
        "location": "Room 302, Engineering Building"
    })
    
    storage.create_study_session({
        "userId": user["id"],
        "title": "Study Group - Algorithms",
        "startTime": today.replace(hour=11, minute=30).isoformat(),
        "endTime": today.replace(hour=13, minute=0).isoformat(),
        "subject": "Computer Science",
        "description": "Group study for algorithm optimization",
        "location": "Library Study Room 4"
    })
    
    storage.create_study_session({
        "userId": user["id"],
        "title": "Assignment Review - AI Ethics",
        "startTime": today.replace(hour=14, minute=0).isoformat(),
        "endTime": today.replace(hour=15, minute=0).isoformat(),
        "subject": "AI Ethics",
        "description": "Review session for the AI ethics assignment",
        "location": "Zoom Meeting"
    })
    
    # Create flashcard sets
    dsa_set = storage.create_flashcard_set({
        "userId": user["id"],
        "title": "Data Structures & Algorithms",
        "description": "Flashcards for DSA course",
        "subject": "Computer Science",
        "tags": ["algorithms", "data structures", "complexity"]
    })
    
    arch_set = storage.create_flashcard_set({
        "userId": user["id"],
        "title": "Computer Architecture",
        "description": "Flashcards for Computer Architecture class",
        "subject": "Computer Science",
        "tags": ["cpu", "memory", "architecture"]
    })
    
    # Create flashcards
    storage.create_flashcard({
        "setId": dsa_set["id"],
        "question": "What is the time complexity of QuickSort in the worst case?",
        "answer": "O(n²) - This occurs when the pivot selection is poor, such as when the array is already sorted.",
        "proficiency": 2
    })
    
    storage.create_flashcard({
        "setId": arch_set["id"],
        "question": "What are the four main components of a CPU?",
        "answer": "1. Arithmetic Logic Unit (ALU)\n2. Control Unit\n3. Registers\n4. Cache",
        "proficiency": 3
    })
    
    # Create study progress entries
    import datetime as dt
    from datetime import date
    from random import randint
    
    for i in range(7):
        # Calculate date for each entry, starting 6 days ago
        entry_date = date.today() - dt.timedelta(days=6-i)
        
        # Randomize study duration between 30 and 240 minutes
        minutes = randint(30, 240)
        
        storage.create_study_progress({
            "userId": user["id"],
            "date": entry_date.isoformat(),
            "studyDuration": minutes,
            "subject": "Various",
            "notes": f"Study session on day {i+1}"
        })

# Initialize the storage with sample data
initialize_storage()